keras/utils/audio_dataset.py

```diff
@@ -0,0 +1,208 @@
+        if class_names:
+            raise ValueError('You can only pass `class_names` if '
+                        f'`labels="inferred"`. Received: labels={labels}, and '
+                        f'class_names={class_names}')
+    if label_mode not in {'int', 'categorical', 'binary', None}:
+        raise ValueError('`label_mode` argument must be one of "int", "categorical", "binary", 'f'or None. Received: label_mode={label_mode}')
+    
+
+
+    if not ragged and output_sequence_length is None:
+        raise Exception(f'The dataset should be ragged dataset or fixed sequence length dataset, found ragged={ragged} and output_sequence_length={output_sequence_length}')
+    elif ragged and output_sequence_length is not None:
+        raise Exception('Cannot set both `ragged` and `output_sequence_length`')
+    
+
+
+    if labels is None or label_mode is None:
+        labels = None
+        label_mode = None
+
+    dataset_utils.check_validation_split_arg(validation_split, subset, shuffle, seed)
+
+    if seed is None:
+        seed = np.random.randint(1e6)
+        
+    file_paths, labels, class_names = dataset_utils.index_directory(directory, labels, formats=ALLOWED_FORMATS, class_names=class_names, 
+        shuffle=shuffle, 
+        seed=seed, 
+        follow_links=follow_links)
+
+    if label_mode == 'binary' and len(class_names) != 2:
+        raise ValueError(f'When passing `label_mode="binary"`, there must be exactly 2 'f'class_names. Received: class_names={class_names}')
+
+    file_paths, labels = dataset_utils.get_training_or_validation_split(file_paths, labels, validation_split, subset)
+    if not file_paths:
+        raise ValueError(f'No audio files found in directory {directory}. 'f'Allowed format(s): {ALLOWED_FORMATS}')
+
+
+    if ragged:
+        dataset = paths_and_labels_to_ragged_dataset(file_paths=file_paths, labels=labels, 
+            label_mode=label_mode, 
+            num_classes=len(class_names), 
+            sampling_rate=sampling_rate)
+    else:
+        dataset = paths_and_labels_to_dataset(file_paths=file_paths, labels=labels, label_mode=label_mode, 
+            num_classes=len(class_names), 
+            sampling_rate=sampling_rate, 
+            output_sequence_length=output_sequence_length)
+
+    dataset = dataset.prefetch(tf.data.AUTOTUNE)
+    if batch_size is not None:
+        if shuffle:
+            dataset = dataset.shuffle(buffer_size=batch_size * 8, seed=seed)
+        dataset = dataset.batch(batch_size)
+    else:
+        if shuffle:
+            dataset = dataset.shuffle(buffer_size=1024, seed=seed)
+
+    # Users may need to reference `class_names`.
+    dataset.class_names = class_names
+    return dataset
+
+
+def prepare_audio(path, sampling_rate, output_sequence_length=-1):
+    """Reads and prepare the audio file."""
+    audio = tf.io.read_file(path)
+    audio, _ = tf.audio.decode_wav(contents=audio, desired_samples=output_sequence_length )
+    if sampling_rate is not None:
```

Comment:
Using encode/decode for resampling is inefficient. We should find a more efficient method to do resampling.

Reply:
Yes I was thinking about that too but at the time I was writing it I could not find any other options but now I found this: https://www.tensorflow.org/io/api_docs/python/tfio/audio/resample

I think it's more efficient but I found that `tensorflow_io` does not get installed by default.

Please let me know if you have other suggestions.

Reply:
A simple solution came to my mind is to try and import `tensorflow_io` and if it exists we will use the `resample` function mentioned above otherwise we will use the current solution and show a warning to the user that recommends installing `tensorflow_io` for more efficient resampling.

@fchollet what do you think ?

Reply:
SGTM, we can use `tensorflow_io` and do a try/except ImportError. Then we can tell people in the error message that they should install tensorflow_io.

The alternatively would have been to add our own utility to resample, but this may be harder than I first thought. In TF-io it seems they write a custom op to do it.

Reply:
`resample` function from tensorflow_io and unit tests are now added.